### PR TITLE
fix(rcm): fix stuck processes on os.fork with gevent

### PR DESCRIPTION
**Pull request by juanjux**
## Description

Fix forked processes getting stuck when running under gevent.

## How

- Reduce the range of the locks in `RemoteConfig`. Since the `RemoteConfigWorker` `start` and `stop` methods already have locks, only protect the assignment to the class member.

- Remove call to `restart` on fork, the `AgentWriter` now takes care of it.

- Implement the `stop_service` and `shutdown` methods for `RemoteConfigWorker` which ensure proper cleanup without zombies. :zombie_man: :zombie_woman: 

- Add a regression test.

## Checklist

- [X] Change(s) are motivated and described in the PR description.
- [X] Testing strategy is described if automated tests are not included in the PR.
- [X] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [X] Change is maintainable (easy to change, telemetry, documentation).
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [ ] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [ ] Title is accurate.
- [ ] No unnecessary changes are introduced.
- [ ] Description motivates each change.
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Testing strategy adequately addresses listed risk(s).
- [ ] Change is maintainable (easy to change, telemetry, documentation).
- [ ] Release note makes sense to a user of the library.
- [ ] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/internal/writer.py [hunk 1]
```diff
@@ -36,6 +37,7 @@
 from .encoding import JSONEncoderV2
 from .encoding import MSGPACK_ENCODERS
 from .logger import get_logger
+from .remoteconfig import RemoteConfig
```
Comment (gnufede) [inline ddtrace/internal/writer.py:40]: nit: duplicated import

File: tests/internal/remoteconfig/test_remoteconfig.py [hunk 2]
```diff
@@ -124,7 +131,7 @@ def test_remote_config_forksafe():
 
         if os.fork() == 0:
             assert RemoteConfig._worker is not None
-            assert RemoteConfig._worker is not parent_worker
+            assert RemoteConfig._worker is parent_worker
```
Comment (emmettbutler) [inline tests/internal/remoteconfig/test_remoteconfig.py:134]: Does this mean that the new intended behavior is for both parent and child processes to share the same `RemoteConfig._worker`?

**Review by emmettbutler — APPROVED** [review]

**Review by emmettbutler — COMMENTED** [review]
What do you think is the explanation for the performance regressions?

File: ddtrace/internal/writer.py [hunk 3]
```diff
@@ -532,6 +533,10 @@ def write(self, spans=None):
                     # are initialized
                     if asbool(os.getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", True)):
                         telemetry_writer.enable()
+
+                    if asbool(os.getenv("DD_REMOTE_CONFIGURATION_ENABLED", True)):
+                        RemoteConfig.enable()
```
Comment (Yun-Kim) [inline ddtrace/internal/writer.py:538]: Can you help me understand how the `AgentWriter` takes care of the call to restart on fork now?

**Review by Yun-Kim — COMMENTED** [review]
Can we get a more detailed description of what the problem was on Gevent? It's not exactly clear why we need to do these things to fix forked processes getting stuck.

I also cannot tell after looking through the changes why we're seeing a significant bump in memory usage (rss in the benchmark scenarios), but it appears consistent enough across most of the benchmark scenarios. I'd like to get some discussion on this before we merge this.

**Comment by juanjux** [comment]
Update: set it to draft for now. Going to investigate if I can find the problem with the forksafe module with gevent since that seems to be where the real problem lies (forksafe = fails, remove it = works).

**Comment by P403n1x87** [comment]
The fix for this issue is essentially in #5105

**Pull request closed without merging**